### PR TITLE
More configs for apprise and more

### PR DESCRIPTION
**Pull request by Guido30**
I've added https://github.com/klept0/MS-Rewards-Farmer/issues/195 and https://github.com/klept0/MS-Rewards-Farmer/issues/194, in the process i realized having both config-private.yaml and config.yaml was unnecessary, so now config.yaml is automatically ignored and only config.yaml.sample should be committed, the user should manually rename the file after configuring it as instructed in the updated README

I was getting the same problem as #185 so i tried using the --disable-http2 flag and it seems to have fixed the issue

i've also kinda fixed the ipapi ratelimited, it will still happen but now it instead of always using US as the default it will use the user configured country code, this does not block searches anymore for that day, the language is useless since its just the browser language, default will still be 'en'

### CODE REVIEW

File: config.yaml.sample
Comment (cal4) [inline config.yaml.sample:1]: IMO this .sample just complicates things needlessly. We should just have the file with defaults. Just isn't a common practice with config (with exception of private files usually).

File: .template-config-private.yaml
Comment (cal4) [inline .template-config-private.yaml:1]: This should be restored, not having it increases the chances someone accidentally commits secret information and from personal experience makes it harder to develop as a result.

File: config.yaml.sample [hunk 1]
```diff
@@ -0,0 +1,13 @@
+# RENAME THIS FILE TO config.yaml ONCE CONFIGURED
+apprise:
+  summary: ON_ERROR
+  exceptions: True  # True or False (Whether to send or not exceptions)
+  promotions: True  # True or False (Whether to send or not incomplete promotions)
```
Comment (cal4) [inline config.yaml.sample:5]: To be a bit clearer, it'd be nice to change this to `apprise.notify.uncaught-exceptions` and `apprise.notify.incomplete-promotions`. Reads a bit better/is clearer IMO.

**Review by cal4 — CHANGES_REQUESTED** [review]
Nice changes!

A few changes that need to be made though

File: .template-config-private.yaml
Comment (cal4) [inline .template-config-private.yaml:1]: This should be restored, not having it increases the chances someone accidentally commits secret information and from personal experience makes it harder to develop as a result.

It's a bit jarring for people to move things back from config-private.yaml to config.yaml again as well.

File: main.py [hunk 2]
```diff
@@ -40,10 +40,11 @@ def main():
             earned_points = executeBot(currentAccount, args)
         except Exception as e1:
             logging.error("", exc_info=True)
-            Utils.sendNotification(
-                f"⚠️ Error executing {currentAccount.username}, please check the log",
-                traceback.format_exc(),
-            )
+            if Utils.loadConfig().get("apprise", {}).get("exceptions", True):
+                Utils.sendNotification(
```
Comment (cal4) [inline main.py:44]: To simplify things, we could add an optional exception parameter (defaults to None). If an exception is provided in sendNotification, we can check the config to see if sending is enabled. If it is, we send. If it isn't, we just return.

**Comment by Guido30** [comment]
I'm still not a fan of using two different config files based on whether they are private values or not, having a .sample file and having to rename its a part of setting up the script, which falls on the user, in the same way that an accounts.json.sample is provided and has to be renamed, why would config be different? the same way where if the accounts.json is not found it will be logged in the console, this could have been done for config.yaml too.

If a default configuration must be provided then an hardcoded one could be used in case the file is missing, this is my opinion.

Anyways i've reverted the changes related to the config and kept everything else in the lastest commit

**Comment by Guido30** [comment]
Another "logical" problem i've found when adding the config values, is when loadConfig is called, opening the file and reading it everytime a single value is needed is not optimal, now the script is small so this doesnt really matter performance wise, but still its not an optimal approach in my opinion, the configuration file should get loaded into memory when the script launches, then values are read when needed from a global variable containing all config values, i wanted to do it this way but the current way that the code is structured doesnt allow it without major refactoring

my two cents

**Comment by cal4** [comment]
> I'm still not a fan of using two different config files based on whether they are private values or not, having a .sample file and having to rename its a part of setting up the script, which falls on the user, in the same way that an accounts.json.sample is provided and has to be renamed, why would config be different? the same way where if the accounts.json is not found it will be logged in the console, this could have been done for config.yaml too.
> 
> If a default configuration must be provided then an hardcoded one could be used in case the file is missing, this is my opinion.
> 
> Anyways i've reverted the changes related to the config and kept everything else in the lastest commit

The general idea is to put secrets/passwords/etc in a file that's gitignored. This prevents accidentally leaking confidential information. One problem that it creates, however, is that it's not exactly obvious what should be in the file that's gitignored (and you can't commit a file, and then gitignore it, once tracked all future changes are tracked). So, the general solution to that is to create a .template. file that can be copied into the new already .gitgnored location.

I didn't give accounts.json + config-private.yaml much thought (honestly forgot until you mentioned), but I think a good practice eventually going forward would be to move accounts.json config to config-private.yaml in the future to make things a bit more straightforward. Having a generalized secrets/private file would be nice.

**Comment by cal4** [comment]
> Another "logical" problem i've found when adding the config values, is when loadConfig is called, opening the file and reading it everytime a single value is needed is not optimal, now the script is small so this doesnt really matter performance wise, but still its not an optimal approach in my opinion, the configuration file should get loaded into memory when the script launches, then values are read when needed from a global variable containing all config values, i wanted to do it this way but the current way that the code is structured doesnt allow it without major refactoring
> 
> my two cents

Yeah, you can use https://github.com/klept0/MS-Rewards-Farmer/blob/0f0052cb705de9752e0fa5e174dda11f2fc300dd/src/utils.py#L38, but not in a static way. Like you're saying though it's probably not a huge performance problem.

File: .template-config-private.yaml [hunk 3]
```diff
@@ -1,5 +1,5 @@
-# config-private.yaml
-# Copy this file to config-private.yaml to use
-apprise:
-  urls:
-    - 'discord://WebhookID/WebhookToken'  # Replace with your actual Apprise service URLs
+# config-private.yaml
+apprise:
+  urls:
+    - "discord://{WebhookID}/{WebhookToken}" # Replace with your actual Apprise service URLs
+default_geolocation: US # Replace with your country code https://en.wikipedia.org/wiki/ISO_3166-1_alpha-2
```
Comment (cal4) [inline .template-config-private.yaml:5]: Probably safe to keep this in config.yaml

**Comment by cal4** [comment]
Good stuff @Guido30, appreciate you fixing the timeout issue

File: src/browser.py [hunk 4]
```diff
@@ -205,8 +206,9 @@ def getCCodeLang(lang: str, geo: str) -> tuple:
             try:
                 nfo = ipapi.location()
             except RateLimited:
-                logging.warning("Returning default", exc_info=True)
-                return "en", "US"
+                geo = Utils.loadConfig("config-private.yaml").get("default_geolocation", "US").upper()
```
Comment (cal4) [inline src/browser.py:209]: If this is explicitly set, I think we should prefer it to the location based on IP.

**Pull request merged**